### PR TITLE
Replace daemon-reload notify with command

### PR DESCRIPTION
**Pull request by fissurezone**

### CODE REVIEW

File: tasks/datanode.yml [hunk 1]
```diff
@@ -8,13 +8,16 @@
...
   when: ansible_service_mgr != "systemd"
 
 - name: Deploy systemd service for datanode
   template: src=hdfs.service.j2 dest=/lib/systemd/system/hdfs-datanode.service owner=root group=root mode=0644
-  notify: Reload systemd daemon
   when: ansible_service_mgr == "systemd"
 
+- name: Reload systemd daemon
+  command: "systemctl daemon-reload"
+  when: ansible_service_mgr == "systemd"
```
Comment (froesef) [inline tasks/datanode.yml]: should this be executed every time or only when the command before changed something?

File: tasks/journalnode.yml [hunk 2]
```diff
@@ -7,13 +7,16 @@
...
   when: ansible_service_mgr != "systemd"
 
 - name: Deploy systemd service for journalnode
   template: src=hdfs.service.j2 dest=/lib/systemd/system/hdfs-journalnode.service owner=root group=root mode=0644
-  notify: Reload systemd daemon
   when: ansible_service_mgr == "systemd"
 
+- name: Reload systemd daemon
+  command: "systemctl daemon-reload"
+  when: ansible_service_mgr == "systemd"
```
Comment (froesef) [inline tasks/journalnode.yml]: should this be executed every time or only when the command before changed something?

File: tasks/datanode.yml (hunk 1 above)
Comment (fissurezone) [inline tasks/datanode.yml]: only required on change, will update

**Comment by fissurezone** [comment]
@froesef updated

**Comment by froesef** [comment]
Now ansible-lint is failing :-(.
You can add the following tag to the action in order to ignore it:
```
tags:
- skip_ansible_lint
```

File: tasks/namenode.yml [hunk 3]
```diff
@@ -18,31 +18,34 @@
...
   when: hdfs_ha_enabled and ansible_service_mgr != "systemd"
 
 - name: Deploy systemd service for zkfc
   template: src=hdfs.service.j2 dest=/lib/systemd/system/hdfs-zkfc.service owner=root group=root mode=0644
-  notify: Reload systemd daemon
   when: hdfs_ha_enabled and ansible_service_mgr == "systemd"
 
+- name: Reload systemd daemon
+  command: "systemctl daemon-reload"
+  when: ansible_service_mgr == "systemd" and ( namenode_systemd_unit_altered.changed or zkfc_systemd_unit_altered.changed )
```
Comment (lhoss) [inline tasks/namenode.yml]: I want us to migrate to using multiline 'when' .. esp. so much nicer in this case
( but ok to leave now .. as the whole role has a number of when's .. many to change)

**Comment by vincepii** [comment]
> You can add the following tag to the action in order to ignore it:
> 
> tags:
> - skip_ansible_lint

And maybe a comment saying that this cannot be relayed to the handlers because the service is needed before the end of the plays!

File: tasks/datanode.yml [hunk 4]
```diff
@@ -8,13 +8,17 @@
...
 - name: Deploy systemd service for datanode
   template: src=hdfs.service.j2 dest=/lib/systemd/system/hdfs-datanode.service owner=root group=root mode=0644
-  notify: Reload systemd daemon
+  register: datanode_systemd_unit_altered
   when: ansible_service_mgr == "systemd"
 
+- name: Reload systemd daemon
+  command: "systemctl daemon-reload"
+  when: ansible_service_mgr == "systemd" and datanode_systemd_unit_altered.changed
+  changed_when: false
```
Comment (lhoss) [inline tasks/datanode.yml]: @fissurezone plz explain why you want to set 'changed_when: false'  (in all reload tasks..) ?
( I assume there won't be any change, WHEN the task is skipped iff `when` evals to false  (when for ex. `datanode_systemd_unit_altered.changed == false` )

**Comment by lhoss** [comment]
LGTM/approved after:
* skip ansible_lint  (see @vincepii comment)
* changed_when  is fixed (or I am proven wrong)

**Comment by Rudd-O** [comment]
THere's a better fix... stand by.

**Comment by Rudd-O** [comment]
The service module in Ansible 2.2 supports this:

```
  - service: name=upower state=started daemon_reload={{ omit if ansible_service_mgr!="systemd" else "yes" }}

```

The parameter is omitted when systemd is not available.  So this works.

Now we can return to the proper handler use.

Enjoy!

**Comment by lhoss** [comment]
> The service module in Ansible 2.2 supports this:

Nice @Rudd-O .. for beeing persistent!
I saw also an open-source PR where somebody used the attribute, on the (old) *service* module .. but gave up on it, after it was not documented in : https://docs.ansible.com/ansible/service_module.html
( so I thought it was only supported by the new/v2.2 *systemd* module ! )

**Comment by fissurezone** [comment]
@lhoss @Rudd-O I am stuck at ansible 2.1. should I introduce the `daemon_reload w/ omit` change?

**Comment by fissurezone** [comment]
@froesef @vincepii is `skip_ansible_lint` preferable to dropping the conditional from the template change?

**Comment by lhoss** [comment]
>  I am stuck at ansible 2.1.
any (major) reason you can't update? 

Would be nice to use @Rudd-O 's  fix .. it's just that if we still want to support ansible <v2.2 , we still have to keep the extra reload task  (ok, at some point we can def. remove pre v2.2 support, and then it looks nice)

**Comment by lhoss** [comment]
> @froesef @vincepii is skip_ansible_lint preferable to dropping the conditional from the template change?

+1 from my side

**Comment by fissurezone** [comment]
@lhoss you will see random bits of roles failing when you deploy m1 prod. Example: bugi.users

2.2 is enforcing on unquoted references, but terrible with error messages from such

**Comment by lhoss** [comment]
> 2.2 is enforcing on unquoted references, but terrible with error messages from such

that's what I initally thought (why the role might fail .. but the errors we got  were indeed misleading)
Anyway @vincepii , I'ld say another TODO (urgent) Story for coming sprint, make our (required)roles (like this bugi) compatible with ansible v2.2 .. CC @Rudd-O

File: tasks/datanode.yml (hunk 4 above)
Comment (fissurezone) [inline tasks/datanode.yml]: `systemctl daemon-reload` is not idempotent (no matter what you do, there may be some difference in the output). `changed_when: false` was to ignore such differences (for ansible lint). Of course, this was from before` skip_ansible_lint`. dropped now in it's favour

**Comment by fissurezone** [comment]
@lhoss @froesef @Rudd-O review updates

**Comment by fissurezone** [comment]
@lhoss @vincepii absolutely forgot about this PR! piggy backing the skip cert checking into this one (needed to change sg download url to https (to work around crazy m1 firewall rules, and apache.org's broken cert file))

**Comment by fissurezone** [comment]
merge without travis?

**Comment by lhoss** [comment]
@fissurezone 
about the broken travis/molecule tests, probably due to newer molecule getting fetched 
( I'ld try with hardcoding to older v1.19, till we have time to make it work with latest version)

**Comment by fissurezone** [comment]
@lhoss can you make that change? and then merge this

**Comment by lhoss** [comment]
@fissurezone travis fixed :)

**Comment by fissurezone** [comment]
thanks @lhoss :) 🥇

**Pull request merged**